File: dart/3.longest_substring.dart
import 'dart:math';

//brute force
// int lengthOfLongestSubstring(String s) {
//   if (s.length == 1) return 1;
//   if (s.length == 0) return 0;
//   int maxLength = 0;
//   for (int i = 0; i < s.length; i++) {
//     String subString = "";
//     for (int j = i; j < s.length; j++) {
//       subString = subString + s[j];
//       if (hasUnique(subString)) {
//         maxLength = max(maxLength, subString.length);
//       }
//     }
//   }
//   return maxLength;
// }

//sliding window
int lengthOfLongestSubstring(String s) {
  if (s.isEmpty) return 0;
  if (s.length == 1) return 1;
  int maxLength = 0;
  int start = 0;
  int end = 0;
  Set chars = {};
  while (end < s.length) {
    if (!chars.contains(s[end])) {
      chars.add(s[end]);
      end++;
    } else {
      chars.remove(s[start]);
      start++;
    }
    maxLength = max(maxLength, chars.length);
  }
  return maxLength;
}

bool hasUnique(String s) {
  return s.split("").toSet().length == s.length;
}

void main() {
  print(lengthOfLongestSubstring("pwwkew"));
}

File: python/3.longest_substring.py
import datetime
# import operator as op

class Solution(object):

    def hasUniqueChar(self,s:str) ->bool:
        unique=set(s)
        return  len(unique) == len(s)
    
#brute force
    # def lengthOfLongestSubstring(self, s: str) -> int:
    #     length=len(s)
    #     if length ==1:
    #         return 1
    #     if length == 0:
    #         return 0
    #     max_length :int =0
    #     for i in range(length):
    #         for j in range(i,length):
    #             sub_string =s[i:j+1]

    #             if self.hasUniqueChar(sub_string):
    #                 max_length= max(max_length,len(sub_string))
    #     return max_length 
    
#sliding window
    def lengthOfLongestSubstring(self, s: str) -> int:
        s_length=len(s)
        max_length=0
        start =0
        end=0
        sub_set= set({})
        while end < s_length:
            if s[end] not in sub_set:
                sub_set.add(s[end])
                max_length=max(max_length,end-start+1)
                end+=1
                print(f"not in -> {sub_set}")
            else:
                sub_set.remove(s[start])
                start+=1
                print(f"in -> {sub_set}")

        return max_length



solution=Solution()
#print(datetime.datetime.now())
print(solution.lengthOfLongestSubstring("pwwkew"))
#print(datetime.datetime.now())


        